src/py21cmfast/inputs.py

```diff
@@ -931,3 +918,47 @@ def t_STAR(self):
             raise ValueError("t_STAR must be above zero and less than or equal to one")
         else:
             return self._t_STAR
+
+
+class InputCrossValidationError(ValueError):
+    """Error when two parameters from different structs aren't consistent."""
+
+    pass
+
+
+def validate_all_inputs(
+    user_params: UserParams,
+    cosmo_params: CosmoParams,
+    astro_params: AstroParams | None = None,
+    flag_options: FlagOptions | None = None,
+):
+    """Cross-validate input parameters from different structs.
+
+    The input params may be modified in-place in this function, but if so, a warning
+    should be emitted.
+    """
+    if astro_params is not None:
+        if astro_params.R_BUBBLE_MAX > user_params.BOX_LEN:
+            astro_params.update(R_BUBBLE_MAX=user_params.BOX_LEN)
+            warnings.warn(
+                f"Setting R_BUBBLE_MAX to BOX_LEN (={user_params.BOX_LEN} as it doesn't make sense for it to be larger."
+            )
+
+        if (
+            global_params.HII_FILTER == 1
+            and astro_params.R_BUBBLE_MAX < user_params.BOX_LEN / 3
```

Comment:
How does this behave with the inhomogeneous recombinations condition (that is, R_BUBBLE_MAX = 50 if set to true)? Does some weirdness occur here whereby R_BUBBLE_MAX != 50 if INHOMO_RECO = True?

Just thinking that in the case of INHOMO_RECO = True and BOX_LEN = 125 the result will be R_BUBBLE_MAX < 50. Not sure if this is desirable... Not sure if twice the size instead of three times the size will help mitigate this.

Reply:
@BradGreig, these conditions don't reference `INHOMO_RECO` at all, so they'll execute regardless of its value. Nevertheless, if `R_BUBBLE_MAX = 50` and `BOX_LEN=125`, nothing should happen to `R_BUBBLE_MAX` here. It's only if `R_BUBBLE_MAX` was smaller than 42 that it would set it back up to ~42. 

I actually wasn't perfectly sure of the correct logic here (physically speaking). I was just following the bare bones of what I wrote in #112. What do you think the logic should be?

One more thing: auto-setting the `R_BUBBLE_MAX` like we do here is actually changing tests, so I'm wondering if it's a little to restrictive, or whether we should change our tests...